File: src/main/java/com/demonwav/mcdev/MinecraftSettings.java
/*
 * Minecraft Dev for IntelliJ
 *
 * https://minecraftdev.org
 *
 * Copyright (c) 2016 minecraft-dev
 *
 * MIT License
 */

package com.demonwav.mcdev;

import com.intellij.openapi.components.PersistentStateComponent;
import com.intellij.openapi.components.ServiceManager;
import com.intellij.openapi.components.State;
import com.intellij.openapi.components.Storage;
import com.intellij.openapi.components.StoragePathMacros;
import com.intellij.openapi.editor.markup.EffectType;
import org.jetbrains.annotations.Contract;
import org.jetbrains.annotations.NotNull;

@State(
        name = "MinecraftSettings",
        storages = @Storage(file = StoragePathMacros.APP_CONFIG + "/minecraft_dev.xml")
)
public class MinecraftSettings implements PersistentStateComponent<MinecraftSettingsState> {

    @NotNull
    private MinecraftSettingsState state = new MinecraftSettingsState();

    @NotNull
    public static MinecraftSettings getInstance() {
        return ServiceManager.getService(MinecraftSettings.class);
    }

    @NotNull
    @Override
    public MinecraftSettingsState getState() {
        return state;
    }

    @Override
    public void loadState(MinecraftSettingsState state) {
        this.state = state;
    }

    // State mappings
    public boolean isShowProjectPlatformIcons() {
        return state.isShowProjectPlatformIcons();
    }

    public void setShowProjectPlatformIcons(boolean showProjectPlatformIcons) {
        state.setShowProjectPlatformIcons(showProjectPlatformIcons);
    }

    public boolean isShowEventListenerGutterIcons() {
        return state.isShowEventListenerGutterIcons();
    }

    public void setShowEventListenerGutterIcons(boolean showEventListenerGutterIcons) {
        state.setShowEventListenerGutterIcons(showEventListenerGutterIcons);
    }

    public boolean isShowChatColorGutterIcons() {
        return state.isShowChatColorGutterIcons();
    }

    public void setShowChatColorGutterIcons(boolean showChatColorGutterIcons) {
        state.setShowChatColorGutterIcons(showChatColorGutterIcons);
    }

    public boolean isShowChatColorUnderlines() {
        return state.isShowChatColorUnderlines();
    }

    public void setShowChatColorUnderlines(boolean showChatColorUnderlines) {
        state.setShowChatColorUnderlines(showChatColorUnderlines);
    }

    @NotNull
    public UnderlineType getUnderlineType() {
        return state.getUnderlineType();
    }

    public void setUnderlineType(@NotNull UnderlineType underlineType) {
        state.setUnderlineType(underlineType);
    }

    public int getUnderlineTypeIndex() {
        final UnderlineType type = getUnderlineType();
        for (int i = 0; i < UnderlineType.values().length; i++) {
            if (type == UnderlineType.values()[i]) {
                return i;
            }
        }
        return 0;
    }

    public boolean isEnableSideOnlyChecks() {
        return state.isEnableSideOnlyChecks();
    }

    public void setEnableSideOnlyChecks(boolean enableSideOnlyChecks) {
        state.setEnableSideOnlyChecks(enableSideOnlyChecks);
    }

    public enum UnderlineType {

        NORMAL("Normal", EffectType.LINE_UNDERSCORE),
        BOLD("Bold", EffectType.BOLD_LINE_UNDERSCORE),
        DOTTED("Dotted", EffectType.BOLD_DOTTED_LINE),
        BOXED("Boxed", EffectType.BOXED),
        ROUNDED_BOXED("Rounded Boxed", EffectType.ROUNDED_BOX),
        WAVED("Waved", EffectType.WAVE_UNDERSCORE);

        private final String regular;
        private final EffectType effectType;

        UnderlineType(@NotNull final String regular, @NotNull final EffectType effectType) {
            this.regular = regular;
            this.effectType = effectType;
        }

        @NotNull
        @Override
        @Contract(pure = true)
        public String toString() {
            return regular;
        }

        @NotNull
        public EffectType getEffectType() {
            return effectType;
        }
    }
}

File: src/main/kotlin/com/demonwav/mcdev/MinecraftSettings.kt
/*
 * Minecraft Dev for IntelliJ
 *
 * https://minecraftdev.org
 *
 * Copyright (c) 2019 minecraft-dev
 *
 * MIT License
 */

package com.demonwav.mcdev

import com.intellij.openapi.components.PersistentStateComponent
import com.intellij.openapi.components.ServiceManager
import com.intellij.openapi.components.State
import com.intellij.openapi.components.Storage
import com.intellij.openapi.editor.markup.EffectType

@State(name = "MinecraftSettings", storages = [Storage("minecraft_dev.xml")])
class MinecraftSettings : PersistentStateComponent<MinecraftSettings.State> {

    data class State(
        var isShowProjectPlatformIcons: Boolean = true,
        var isShowEventListenerGutterIcons: Boolean = true,
        var isShowChatColorGutterIcons: Boolean = true,
        var isShowChatColorUnderlines: Boolean = false,
        var underlineType: MinecraftSettings.UnderlineType = MinecraftSettings.UnderlineType.DOTTED
    )

    private var state = State()

    override fun getState(): State {
        return state
    }

    override fun loadState(state: State) {
        this.state = state
    }

    // State mappings
    var isShowProjectPlatformIcons: Boolean
        get() = state.isShowProjectPlatformIcons
        set(showProjectPlatformIcons) {
            state.isShowProjectPlatformIcons = showProjectPlatformIcons
        }

    var isShowEventListenerGutterIcons: Boolean
        get() = state.isShowEventListenerGutterIcons
        set(showEventListenerGutterIcons) {
            state.isShowEventListenerGutterIcons = showEventListenerGutterIcons
        }

    var isShowChatColorGutterIcons: Boolean
        get() = state.isShowChatColorGutterIcons
        set(showChatColorGutterIcons) {
            state.isShowChatColorGutterIcons = showChatColorGutterIcons
        }

    var isShowChatColorUnderlines: Boolean
        get() = state.isShowChatColorUnderlines
        set(showChatColorUnderlines) {
            state.isShowChatColorUnderlines = showChatColorUnderlines
        }

    var underlineType: UnderlineType
        get() = state.underlineType
        set(underlineType) {
            state.underlineType = underlineType
        }

    val underlineTypeIndex: Int
        get() {
            val type = underlineType
            return (0 until UnderlineType.values().size).firstOrNull { type == UnderlineType.values()[it] } ?: 0
        }

    enum class UnderlineType(private val regular: String, val effectType: EffectType) {

        NORMAL("Normal", EffectType.LINE_UNDERSCORE),
        BOLD("Bold", EffectType.BOLD_LINE_UNDERSCORE),
        DOTTED("Dotted", EffectType.BOLD_DOTTED_LINE),
        BOXED("Boxed", EffectType.BOXED),
        ROUNDED_BOXED("Rounded Boxed", EffectType.ROUNDED_BOX),
        WAVED("Waved", EffectType.WAVE_UNDERSCORE);

        override fun toString(): String {
            return regular
        }
    }

    companion object {
        val instance: MinecraftSettings
            get() = ServiceManager.getService(MinecraftSettings::class.java)
    }
}
